prov/tcp/src/tcpx_ep.c

```diff
@@ -0,0 +1,769 @@
+	if (ofi_atomic_get32(&ep->ref)) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL, "EP busy\n");
+		return -FI_EBUSY;
+	}
+
+	ofi_close_socket(ep->handle->conn_fd);
+	ofi_endpoint_close(&ep->util_ep);
+
+	free(ep->handle);
+	free(ep);
+	return 0;
+}
+
+static int tcpx_ep_bind(struct fid *fid, struct fid *bfid, uint64_t flags)
+{
+	struct tcpx_ep *tcpx_ep;
+	struct util_eq *eq;
+	struct util_cq *cq;
+	int ret;
+
+	tcpx_ep = container_of(fid, struct tcpx_ep, util_ep.ep_fid.fid);
+
+	switch(bfid->fclass) {
+	case FI_CLASS_EQ:
+		eq = container_of(bfid, struct util_eq, eq_fid.fid);
+		tcpx_ep->util_ep.eq = eq;
+		ofi_atomic_inc32(&eq->ref);
+
+
+		break;
+	case FI_CLASS_CQ:
+		cq = container_of(bfid, struct util_cq, cq_fid.fid);
+		if (flags & FI_SEND) {
+			tcpx_ep->util_ep.tx_cq = cq;
+			ofi_atomic_inc32(&cq->ref);
+
+			ret = fid_list_insert(&cq->ep_list,
+					      &cq->ep_list_lock,
+					      &tcpx_ep->util_ep.ep_fid.fid);
+			if (ret)
+				return ret;
+
+		}
+		if (flags & FI_RECV) {
+			tcpx_ep->util_ep.rx_cq = cq;
+			ofi_atomic_inc32(&cq->ref);
+
+			ret = fid_list_insert(&cq->ep_list,
+					      &cq->ep_list_lock,
+					      &tcpx_ep->util_ep.ep_fid.fid);
+			if (ret)
+				return ret;
+
+		}
+
+		if(!(flags & (FI_SEND | FI_RECV)))
+			return -FI_ENOSYS;
+		break;
+	case FI_CLASS_MR:
+	case FI_CLASS_AV:
+	default:
+		return -FI_ENOSYS;
+	}
+	return FI_SUCCESS;
+}
+
+
+static int tcpx_ep_ctrl(struct fid *fid, int command, void *arg)
+{
+	struct tcpx_ep *ep;
+
+	ep = container_of(fid, struct tcpx_ep, util_ep.ep_fid.fid);
+	switch (command) {
+	case FI_ENABLE:
+		if (!ep->util_ep.rx_cq || !ep->util_ep.tx_cq)
+			return -FI_ENOCQ;
+		break;
+	default:
+		return -FI_ENOSYS;
+	}
+	return 0;
+}
+
+static struct fi_ops tcpx_ep_fi_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = tcpx_ep_close,
+	.bind = tcpx_ep_bind,
+	.control = tcpx_ep_ctrl,
+	.ops_open = fi_no_ops_open,
+};
+
+static struct fi_ops_ep tcpx_ep_ops = {
+	.size = sizeof(struct fi_ops_ep),
+	.cancel = fi_no_cancel,
+	.getopt = fi_no_getopt, //tcpx_getopt,
+	.setopt = fi_no_setopt, //tcpx_setopt,
+	.tx_ctx = fi_no_tx_ctx,
+	.rx_ctx = fi_no_rx_ctx,
+	.rx_size_left = fi_no_rx_size_left,
+	.tx_size_left = fi_no_tx_size_left,
+};
+
+static void tcpx_manual_progress(struct util_ep *util_ep)
+{
+	return;
+}
+
+int tcpx_endpoint(struct fid_domain *domain, struct fi_info *info,
+		  struct fid_ep **ep_fid, void *context)
+{
+	struct tcpx_ep *ep;
+	struct tcpx_conn_handle *handle;
+	int ret;
+
+	ep = calloc(1, sizeof(*ep));
+	if (!ep)
+		return -FI_ENOMEM;
+
+	ret = ofi_endpoint_init(domain, &tcpx_util_prov, info, &ep->util_ep,
+				context, tcpx_manual_progress);
+	if (ret)
+		goto err;
+
+	ep->info = *info;
+	ofi_atomic_initialize32(&ep->ref, 0);
+
+	if (info != NULL && info->handle != NULL){
+		handle = container_of(info->handle,struct tcpx_conn_handle,
+				      handle);
+		ep->handle = handle;
+		fi_fd_nonblock(handle->conn_fd);
```

Comment:
I would look at moving the socket directly into the ep structure, and not a substructure.  The connection request handle can be freed immediately after the connection completes.

We should allocate the socket in this call, and configure it appropriately.